### PR TITLE
api: allow multiple type URLs for xDS

### LINKED ISSUE #10209: hardcoded v2 type urls in xds requests
*Title*: *hardcoded v2 type urls in xds requests*

*Description*:
In order to migrate to v3 API, `loadTypeUrl` functions are used to determine type url in discovery requests. However, I found that some files are still using v2 type url constants in [this file](https://github.com/envoyproxy/envoy/blob/master/source/common/config/resources.h). This means that even envoy is configured to use v3, it will still send requests for v2 resources.

Correct me if there is any magic actually doing the conversion somewhere.

[optional *Relevant Links*:]
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/common/router/scoped_rds.cc#L243
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/common/router/scoped_rds.cc#L258
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/common/upstream/cds_api_impl.cc#L68-L70
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/common/upstream/cluster_manager_impl.cc#L150-L155
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/common/upstream/cluster_manager_impl.cc#L752-L754
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/server/lds_api.cc#L43-L45
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/server/server.cc#L526-L535

### PR DESCRIPTION
**Pull request by Shikugawa**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: allow multiple versions of type URLs. For example, we can set `type.googleapis.com/envoy.api.v2.ScopedRouteConfiguration` and `"type.googleapis.com/envoy.config.route.v3.ScopedRouteConfiguration"` simultaneously.
Risk Level: Mid 
Testing: Unit / Integration
Docs Changes: N/A
Release Notes: N/A
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: source/common/upstream/cluster_manager_impl.cc [hunk 1]
```diff
@@ -143,16 +142,21 @@ void ClusterManagerInitHelper::maybeFinishInitialize() {
...
-      if (cm_.adsMux() == nullptr ||
-          cm_.adsMux()->paused(Config::TypeUrl::get().ClusterLoadAssignment)) {
+      if (cm_.adsMux() == nullptr) {
         initializeSecondaryClusters();
       } else {
-        cm_.adsMux()->pause(Config::TypeUrl::get().ClusterLoadAssignment);
-        Cleanup eds_resume(
-            [this] { cm_.adsMux()->resume(Config::TypeUrl::get().ClusterLoadAssignment); });
+        for (auto&& type_url : target_type_urls) {
+          if (!cm_.adsMux()->paused(type_url)) {
```
Comment (htuch) [inline source/common/upstream/cluster_manager_impl.cc:155]: FWIW, I think we're still mixing two unrelated PRs; switching to non-hardcoded type URLs and supporting multiple type URLs.

File: source/common/upstream/cluster_manager_impl.cc [hunk 2]
```diff
@@ -143,16 +142,21 @@ void ClusterManagerInitHelper::maybeFinishInitialize() {
   if (!secondary_init_clusters_.empty()) {
     if (!started_secondary_initialize_) {
       ENVOY_LOG(info, "cm init: initializing secondary clusters");
+      const auto target_type_urls =
+          Config::getAllVersionTypeUrls<envoy::config::endpoint::v3::ClusterLoadAssignment>();
```
Comment (htuch) [inline source/common/upstream/cluster_manager_impl.cc:146]: This seems a regression in terms of reducing the amount of version hardcoding; before we could just do `Config::TypeUrl::get().ClusterLoadAssignment()`, now we need to mention v3..

File: include/envoy/config/grpc_mux.h [hunk 3]
```diff
@@ -63,13 +65,36 @@ class GrpcMux {
    */
   virtual void pause(const std::string& type_url) PURE;
 
+  /**
+   * Pause discovery requests for all of given API types. This is useful when we're processing an
+   * update for LDS or CDS and don't want a flood of updates for RDS or EDS respectively. Discovery
+   * requests may later be resumed with resume().
+   * @param type_urls all of type URLs corresponding to xDS APIs.
+   */
+  void pause(const std::vector<std::string>& type_urls) {
```
Comment (htuch) [inline include/envoy/config/grpc_mux.h:74]: Do we need some unit tests for these vectored pause/resume methods?

**Comment by Shikugawa** [comment]
@htuch Thanks. I separated this PR into this and #10848. When finished #10848, I will continue to work this to allow multiple type URLs for the same resources.

**Pull request closed without merging**